File: BuddyGrammarKit/Sources/BuddyGrammarKit/SwipeTypingEngine.swift
import CoreGraphics
import Foundation

/// QWERTY key centers in a normalized coordinate space where adjacent keys in
/// a row are 1 apart and rows are 1 apart, with the middle and bottom rows
/// offset like a physical keyboard.
public enum QwertyKeyLayout {
    public static let neighborDistance = 1.3

    public static func position(of character: Character) -> (x: Double, y: Double)? {
        positions[Character(character.lowercased())]
    }

    public static func center(of character: Character) -> CGPoint? {
        position(of: character).map { CGPoint(x: $0.x, y: $0.y) }
    }

    public static func distance(_ lhs: Character, _ rhs: Character) -> Double {
        guard let a = position(of: lhs), let b = position(of: rhs) else {
            return .infinity
        }
        return hypot(a.x - b.x, a.y - b.y)
    }

    static func nearestKey(to point: CGPoint, maximumDistance: Double = 0.8) -> Character? {
        positions
            .map { key, center in
                (key, hypot(center.x - point.x, center.y - point.y))
            }
            .filter { $0.1 <= maximumDistance }
            .min { $0.1 < $1.1 }?
            .0
    }

    private static let positions: [Character: (x: Double, y: Double)] = {
        let rows: [(String, Double)] = [
            ("qwertyuiop", 0),
            ("asdfghjkl", 0.25),
            ("zxcvbnm", 0.75),
        ]
        var positions: [Character: (x: Double, y: Double)] = [:]
        for (rowIndex, row) in rows.enumerated() {
            for (columnIndex, character) in row.0.enumerated() {
                positions[character] = (Double(columnIndex) + row.1, Double(rowIndex))
            }
        }
        return positions
    }()
}

/// A monotonic swipe sample in normalized key-space. Milliseconds are used
/// explicitly so shared Swift/Kotlin conformance traces have the same unit.
public struct SwipePathSample: Equatable, Sendable {
    public let x: Double
    public let y: Double
    public let timestampMilliseconds: Double

    public init(x: Double, y: Double, timestampMilliseconds: Double) {
        self.x = x
        self.y = y
        self.timestampMilliseconds = timestampMilliseconds
    }

    public init(point: CGPoint, timestampMilliseconds: Double) {
        self.init(
            x: point.x,
            y: point.y,
            timestampMilliseconds: timestampMilliseconds
        )
    }

    fileprivate var point: CGPoint { CGPoint(x: x, y: y) }
}

public struct SwipeDwellConfiguration: Equatable, Sendable {
    public let minimumMilliseconds: Double
    public let minimumSamples: Int
    public let maximumDriftKeyUnits: Double

    public init(
        minimumMilliseconds: Double,
        minimumSamples: Int,
        maximumDriftKeyUnits: Double
    ) {
        self.minimumMilliseconds = minimumMilliseconds
        self.minimumSamples = minimumSamples
        self.maximumDriftKeyUnits = maximumDriftKeyUnits
    }

    /// Mirrors keyboard catalog v1. Native catalog loaders can inject a newer
    /// validated configuration without changing recognition callers.
    public static let contractV1 = Self(
        minimumMilliseconds: 180,
        minimumSamples: 3,
        maximumDriftKeyUnits: 0.42
    )
}

public struct SwipeCandidate: Equatable, Sendable {
    public let word: String
    public let confidence: Double

    public init(word: String, confidence: Double) {
        self.word = word
        self.confidence = confidence
    }
}

public enum SwipeAbstentionReason: String, Equatable, Sendable {
    case tooShort
    case invalidSamples
    case noCandidate
    case lowConfidence
    case ambiguous
}

public struct SwipeRecognitionResult: Equatable, Sendable {
    public let candidates: [SwipeCandidate]
    public let acceptedCandidate: SwipeCandidate?
    public let confidence: Double
    public let margin: Double
    public let abstentionReason: SwipeAbstentionReason?

    public var abstained: Bool { acceptedCandidate == nil }

    fileprivate init(
        candidates: [SwipeCandidate],
        acceptedCandidate: SwipeCandidate?,
        confidence: Double,
        margin: Double,
        abstentionReason: SwipeAbstentionReason?
    ) {
        self.candidates = candidates
        self.acceptedCandidate = acceptedCandidate
        self.confidence = confidence
        self.margin = margin
        self.abstentionReason = abstentionReason
    }

    fileprivate static func abstaining(_ reason: SwipeAbstentionReason) -> Self {
        Self(
            candidates: [],
            acceptedCandidate: nil,
            confidence: 0,
            margin: 0,
            abstentionReason: reason
        )
    }
}

/// SHARK²-style gesture recognizer for swipe typing.
///
/// The finger path (in key-space coordinates) is resampled to a fixed number
/// of equidistant points and compared against each candidate word's ideal
/// path through its key centers on two channels: absolute location, and
/// translation/scale-invariant shape. Channel distances are blended with a
/// word-frequency prior and an optional previous-word context boost.
public struct SwipeTypingEngine: Sendable {
    private struct Entry: Sendable {
        let word: String
        let rank: Int
        let languageCode: String?
        let centers: [CGPoint]
        let repeatedLetters: [Character: Int]
        let pathLength: Double
        let firstCenter: CGPoint
        let lastCenter: CGPoint
    }

    private struct WordSource {
        let word: String
        let rank: Int
        let languageCode: String?
    }

    private struct ScoredEntry {
        let word: String
        let score: Double
    }

    private let entries: [Entry]
    private let dwellConfiguration: SwipeDwellConfiguration
    private static let sampleCount = 32
    private static let anchorTolerance = 1.6
    private static let rejectionScore = 0.62
    private static let recognitionScoreLimit = 0.78
    private static let minimumConfidence = 0.50
    private static let minimumMargin = 0.03

    public init(
        extraWords: [String] = [],
        dwellConfiguration: SwipeDwellConfiguration = .contractV1
    ) {
        let lexicon = WordFrequencyLexicon.shared
        var sources: [WordSource] = []
        for languageCode in ["en", "it"] {
            sources.append(contentsOf: lexicon.words(languageCode: languageCode).enumerated().map {
                WordSource(
                    word: $0.element,
                    rank: $0.offset,
                    languageCode: languageCode
                )
            })
        }
        var seen = Set(sources.map(\.word))
        for extra in extraWords {
            guard let normalized = SwipeWordNormalizer.normalize(extra),
                  seen.insert(normalized.display).inserted else { continue }
            sources.append(
                WordSource(word: normalized.display, rank: 400, languageCode: nil)
            )
        }
        self.init(sources: sources, dwellConfiguration: dwellConfiguration)
    }

    /// Deterministic lexicon initializer used by language-pack loaders and
    /// cross-platform conformance tests. `words` are language-neutral while
    /// `languageWords` are considered only for the requested base language.
    public init(
        words: [String],
        languageWords: [String: [String]] = [:],
        dwellConfiguration: SwipeDwellConfiguration = .contractV1
    ) {
        var sources = words.enumerated().map {
            WordSource(word: $0.element, rank: $0.offset, languageCode: nil)
        }
        for languageCode in languageWords.keys.sorted() {
            let baseRank = sources.count
            sources.append(contentsOf: (languageWords[languageCode] ?? []).enumerated().map {
                WordSource(
                    word: $0.element,
                    rank: baseRank + $0.offset,
                    languageCode: Self.baseLanguage(languageCode)
                )
            })
        }
        self.init(sources: sources, dwellConfiguration: dwellConfiguration)
    }

    private init(
        sources: [WordSource],
        dwellConfiguration: SwipeDwellConfiguration
    ) {
        self.dwellConfiguration = dwellConfiguration
        var seen = Set<String>()
        entries = sources.compactMap { source in
            guard let form = SwipeWordNormalizer.normalize(source.word),
                  !form.geometry.isEmpty,
                  form.geometry.allSatisfy({ QwertyKeyLayout.position(of: $0) != nil }) else {
                return nil
            }
            let word = form.display
            guard seen.insert("\(source.languageCode ?? "*")|\(word)").inserted else {
                return nil
            }
            var centers: [CGPoint] = []
            var repeatedLetters: [Character: Int] = [:]
            var lastLetter: Character?
            for letter in form.geometry where letter != lastLetter {
                guard let center = QwertyKeyLayout.center(of: letter) else { return nil }
                centers.append(center)
                lastLetter = letter
            }
            lastLetter = nil
            for letter in form.geometry {
                if letter == lastLetter {
                    repeatedLetters[letter, default: 0] += 1
                }
                lastLetter = letter
            }
            guard let first = centers.first, let last = centers.last else { return nil }
            return Entry(
                word: word,
                rank: source.rank,
                languageCode: source.languageCode,
                centers: centers,
                repeatedLetters: repeatedLetters,
                pathLength: Self.length(of: centers),
                firstCenter: first,
                lastCenter: last
            )
        }
    }

    /// Convenience for tests and callers that only have a key trace: builds a
    /// path through the traced keys' centers.
    public func candidates(
        for trace: [Character],
        limit: Int = 3,
        previousWord: String? = nil
    ) -> [String] {
        let path = trace.compactMap(QwertyKeyLayout.center(of:))
        return candidates(forKeySpacePath: path, limit: limit, previousWord: previousWord)
    }

    /// Returns the best-matching words for a finger path expressed in
    /// key-space coordinates (1 unit = 1 key width), most likely first.
    public func candidates(
        forKeySpacePath path: [CGPoint],
        limit: Int = 3,
        previousWord: String? = nil
    ) -> [String] {
        scoredCandidates(
            path: path,
            repeatedLetters: nil,
            limit: limit,
            previousWord: previousWord,
            // This legacy geometry-only API historically used the bundled
            // English list. Locale-aware production callers use `recognize`.
            languageCode: "en"
        )
            .filter { $0.score <= Self.rejectionScore }
            .map(\.word)
    }

    /// Rich timed recognition Interface. Candidate confidences and the top-two
    /// margin are always returned when ranking succeeds; callers commit only
    /// `acceptedCandidate`, preserving conservative abstention.
    public func recognize(
        samples: [SwipePathSample],
        limit: Int = 3,
        previousWord: String? = nil,
        languageCode: String? = nil
    ) -> SwipeRecognitionResult {
        guard samples.count >= 2, limit > 0 else {
            return .abstaining(.tooShort)
        }
        guard Self.valid(samples: samples) else {
            return .abstaining(.invalidSamples)
        }

        let path = samples.map(\.point)
        let repeatedLetters = Self.repeatedLetterEvidence(
            in: samples,
            configuration: dwellConfiguration
        )
        let scored = scoredCandidates(
            path: path,
            repeatedLetters: repeatedLetters,
            limit: limit,
            previousWord: previousWord,
            languageCode: languageCode
        ).filter { $0.score <= Self.recognitionScoreLimit }
        guard !scored.isEmpty else {
            return .abstaining(.noCandidate)
        }

        let candidates = scored.map {
            SwipeCandidate(word: $0.word, confidence: Self.confidence(for: $0.score))
        }
        let confidence = candidates[0].confidence
        let margin = candidates.count > 1
            ? max(0, confidence - candidates[1].confidence)
            : confidence
        let reason: SwipeAbstentionReason?
        if confidence < Self.minimumConfidence {
            reason = .lowConfidence
        } else if margin < Self.minimumMargin {
            reason = .ambiguous
        } else {
            reason = nil
        }
        return SwipeRecognitionResult(
            candidates: candidates,
            acceptedCandidate: reason == nil ? candidates.first : nil,
            confidence: confidence,
            margin: margin,
            abstentionReason: reason
        )
    }

    private func scoredCandidates(
        path: [CGPoint],
        repeatedLetters: [Character: Int]?,
        limit: Int,
        previousWord: String?,
        languageCode: String?
    ) -> [ScoredEntry] {
        guard path.count >= 2, limit > 0,
              let start = path.first, let end = path.last else {
            return []
        }

        let requestedLanguage = languageCode.map(Self.baseLanguage)
        let eligibleEntries = entries.filter { entry in
            guard let requestedLanguage else { return true }
            return entry.languageCode == nil || entry.languageCode == requestedLanguage
        }
        let sampledPath = Self.resample(path, count: Self.sampleCount)
        let sampledShape = Self.shapeNormalized(sampledPath)
        let pathLength = Self.length(of: path)
        let vocabularySize = max(eligibleEntries.count, 1)
        let continuations = previousWord
            .flatMap { NextWordPredictor.bigrams[$0.lowercased()] } ?? []

        var bestByWord: [String: Double] = [:]
        for entry in eligibleEntries {
            let startDistance = Self.distance(entry.firstCenter, start)
            let endDistance = Self.distance(entry.lastCenter, end)
            guard startDistance <= Self.anchorTolerance,
                  endDistance <= Self.anchorTolerance,
                  abs(log((entry.pathLength + 0.5) / (pathLength + 0.5))) <= log(2.3) else {
                continue
            }

            let idealPath = Self.resample(entry.centers, count: Self.sampleCount)
            let location = Self.meanDistance(sampledPath, idealPath) / 3
            let shape = Self.meanDistance(sampledShape, Self.shapeNormalized(idealPath)) * 2
            let rankScore = log(1 + Double(entry.rank)) / log(1 + Double(vocabularySize))

            var score = 0.40 * min(location, 1.5)
                + 0.32 * min(shape, 1.5)
                + 0.18 * rankScore
                + 0.05 * (startDistance + endDistance)
            if let repeatedLetters {
                score += Self.repeatedLetterScore(
                    expected: entry.repeatedLetters,
                    observed: repeatedLetters
                )
            }
            if continuations.contains(entry.word) {
                score -= 0.10
            }
            bestByWord[entry.word] = min(bestByWord[entry.word] ?? .infinity, score)
        }

        let scored = bestByWord.map {
            ScoredEntry(word: $0.key, score: $0.value)
        }
        let sorted = scored.sorted { lhs, rhs in
            lhs.score == rhs.score ? lhs.word < rhs.word : lhs.score < rhs.score
        }
        return Array(sorted.prefix(limit))
    }

    private static func repeatedLetterScore(
        expected: [Character: Int],
        observed: [Character: Int]
    ) -> Double {
        var score = 0.0
        for key in Set(expected.keys).union(observed.keys) {
            let expectedCount = expected[key, default: 0]
            let observedCount = observed[key, default: 0]
            let matched = min(expectedCount, observedCount)
            let missing = max(0, expectedCount - observedCount)
            let unexpected = max(0, observedCount - expectedCount)
            score -= 0.12 * Double(matched)
            score += 0.18 * Double(missing)
            score += 0.16 * Double(unexpected)
        }
        return score
    }

    private static func confidence(for score: Double) -> Double {
        let value = 1 / (1 + exp(6 * (score - 0.45)))
        return min(1, max(0, value))
    }

    private static func baseLanguage(_ languageCode: String) -> String {
        languageCode
            .lowercased()
            .split(whereSeparator: { $0 == "-" || $0 == "_" })
            .first
            .map(String.init) ?? languageCode.lowercased()
    }

    private static func valid(samples: [SwipePathSample]) -> Bool {
        var previousTimestamp = -Double.infinity
        for sample in samples {
            guard sample.x.isFinite,
                  sample.y.isFinite,
                  sample.timestampMilliseconds.isFinite,
                  sample.timestampMilliseconds >= previousTimestamp else {
                return false
            }
            previousTimestamp = sample.timestampMilliseconds
        }
        return true
    }

    private static func repeatedLetterEvidence(
        in samples: [SwipePathSample],
        configuration: SwipeDwellConfiguration
    ) -> [Character: Int] {
        var runs: [(key: Character, samples: [SwipePathSample])] = []
        for sample in samples {
            guard let key = QwertyKeyLayout.nearestKey(to: sample.point) else { continue }
            if runs.last?.key == key {
                runs[runs.count - 1].samples.append(sample)
            } else {
                runs.append((key, [sample]))
            }
        }

        var evidence: [Character: Int] = [:]
        for run in runs {
            guard let first = run.samples.first, let last = run.samples.last,
                  run.samples.count >= configuration.minimumSamples,
                  last.timestampMilliseconds - first.timestampMilliseconds
                    >= configuration.minimumMilliseconds else {
                continue
            }
            let maximumDrift = run.samples.map {
                hypot($0.x - first.x, $0.y - first.y)
            }.max() ?? .infinity
            guard maximumDrift <= configuration.maximumDriftKeyUnits else { continue }
            evidence[run.key, default: 0] += 1
        }
        return evidence
    }

    // MARK: - Geometry

    private static func distance(_ lhs: CGPoint, _ rhs: CGPoint) -> Double {
        hypot(lhs.x - rhs.x, lhs.y - rhs.y)
    }

    private static func length(of points: [CGPoint]) -> Double {
        guard points.count > 1 else { return 0 }
        var total = 0.0
        for index in 1..<points.count {
            total += distance(points[index - 1], points[index])
        }
        return total
    }

    /// Resamples a polyline into `count` equidistant points.
    private static func resample(_ points: [CGPoint], count: Int) -> [CGPoint] {
        guard let first = points.first else { return [] }
        let total = length(of: points)
        guard points.count > 1, total > 0 else {
            return Array(repeating: first, count: count)
        }

        let interval = total / Double(count - 1)
        var result: [CGPoint] = [first]
        var accumulated = 0.0
        var previous = first

        for point in points.dropFirst() {
            var segment = distance(previous, point)
            var segmentStart = previous
            while accumulated + segment >= interval, result.count < count {
                let ratio = (interval - accumulated) / segment
                let sample = CGPoint(
                    x: segmentStart.x + ratio * (point.x - segmentStart.x),
                    y: segmentStart.y + ratio * (point.y - segmentStart.y)
                )
                result.append(sample)
                segment = accumulated + segment - interval
                accumulated = 0
                segmentStart = sample
            }
            accumulated += segment
            previous = point
        }

        while result.count < count {
            result.append(points[points.count - 1])
        }
        return result
    }

    /// Translation- and scale-invariant form of a path: centered on its
    /// centroid and scaled by its bounding box.
    private static func shapeNormalized(_ points: [CGPoint]) -> [CGPoint] {
        guard !points.isEmpty else { return points }
        let count = Double(points.count)
        let centroid = CGPoint(
            x: points.reduce(0) { $0 + $1.x } / count,
            y: points.reduce(0) { $0 + $1.y } / count
        )
        let minX = points.map(\.x).min() ?? 0
        let maxX = points.map(\.x).max() ?? 0
        let minY = points.map(\.y).min() ?? 0
        let maxY = points.map(\.y).max() ?? 0
        let scale = max(maxX - minX, maxY - minY, 0.01)
        return points.map {
            CGPoint(x: ($0.x - centroid.x) / scale, y: ($0.y - centroid.y) / scale)
        }
    }

    private static func meanDistance(_ lhs: [CGPoint], _ rhs: [CGPoint]) -> Double {
        guard lhs.count == rhs.count, !lhs.isEmpty else { return .infinity }
        var total = 0.0
        for index in lhs.indices {
            total += distance(lhs[index], rhs[index])
        }
        return total / Double(lhs.count)
    }
}

File: android/app/src/main/java/com/francescooddo/buddygrammar/core/SwipeTypingEngine.kt
package com.francescooddo.buddygrammar.core

import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.hypot
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min

/** A point in normalized QWERTY key-space, where adjacent keys are one unit apart. */
data class SwipePoint(val x: Double, val y: Double)

/** Monotonic timed sample; milliseconds match the shared Swift/Kotlin traces. */
data class SwipePathSample(
    val x: Double,
    val y: Double,
    val timestampMilliseconds: Double,
) {
    val point: SwipePoint get() = SwipePoint(x, y)
}

data class SwipeDwellConfiguration(
    val minimumMilliseconds: Double,
    val minimumSamples: Int,
    val maximumDriftKeyUnits: Double,
) {
    companion object {
        /** Mirrors the validated keyboard catalog v1 values. */
        val CONTRACT_V1 = SwipeDwellConfiguration(
            minimumMilliseconds = 180.0,
            minimumSamples = 3,
            maximumDriftKeyUnits = 0.42,
        )
    }
}

data class SwipeCandidate(
    val word: String,
    val confidence: Double,
)

enum class SwipeAbstentionReason {
    TOO_SHORT,
    INVALID_SAMPLES,
    NO_CANDIDATE,
    LOW_CONFIDENCE,
    AMBIGUOUS,
}

data class SwipeRecognitionResult(
    val candidates: List<SwipeCandidate>,
    val acceptedCandidate: SwipeCandidate?,
    val confidence: Double,
    val margin: Double,
    val abstentionReason: SwipeAbstentionReason?,
) {
    val abstained: Boolean get() = acceptedCandidate == null

    companion object {
        fun abstaining(reason: SwipeAbstentionReason) = SwipeRecognitionResult(
            candidates = emptyList(),
            acceptedCandidate = null,
            confidence = 0.0,
            margin = 0.0,
            abstentionReason = reason,
        )
    }
}

/** Fixed QWERTY key centers shared by gesture capture and recognition. */
object QwertyKeyLayout {
    private val positions: Map<Char, SwipePoint> = buildMap {
        listOf(
            Triple("qwertyuiop", 0.0, 0.0),
            Triple("asdfghjkl", 0.25, 1.0),
            Triple("zxcvbnm", 0.75, 2.0),
        ).forEach { (row, offset, y) ->
            row.forEachIndexed { index, character ->
                put(character, SwipePoint(index + offset, y))
            }
        }
    }

    fun center(character: Char): SwipePoint? = positions[character.lowercaseChar()]

    fun nearestKey(point: SwipePoint, maximumDistance: Double = 0.8): Char? = positions
        .map { (key, center) -> key to hypot(center.x - point.x, center.y - point.y) }
        .filter { (_, distance) -> distance <= maximumDistance }
        .minByOrNull { (_, distance) -> distance }
        ?.first
}

/**
 * SHARK²-style swipe recognizer matching the iOS keyboard implementation.
 *
 * The captured gesture and each candidate word are resampled into equal-length
 * paths. Absolute location, translation/scale-independent shape, word rank,
 * and sentence context are blended into one rejection score.
 */
class SwipeTypingEngine(
    words: List<String> = WordList.words,
    languageWords: Map<String, List<String>> = emptyMap(),
    private val dwellConfiguration: SwipeDwellConfiguration = SwipeDwellConfiguration.CONTRACT_V1,
) {
    private data class Entry(
        val word: String,
        val geometry: String,
        val rank: Int,
        val languageTag: String?,
        val centers: List<SwipePoint>,
        val repeatedLetters: Map<Char, Int>,
        val pathLength: Double,
        val firstCenter: SwipePoint,
        val lastCenter: SwipePoint,
    )

    private data class WordSource(
        val word: String,
        val rank: Int,
        val languageTag: String?,
    )

    private data class ScoredEntry(
        val word: String,
        val score: Double,
    )

    private data class KeyRun(
        val key: Char,
        val samples: MutableList<SwipePathSample>,
    )

    private val entries: List<Entry>

    init {
        val sources = words.mapIndexed { rank, word ->
            WordSource(word = word, rank = rank, languageTag = null)
        }.toMutableList()
        languageWords.toSortedMap().forEach { (languageTag, taggedWords) ->
            val baseRank = sources.size
            taggedWords.forEachIndexed { index, word ->
                sources += WordSource(
                    word = word,
                    rank = baseRank + index,
                    languageTag = baseLanguage(languageTag),
                )
            }
        }

        val seen = mutableSetOf<String>()
        entries = sources.mapNotNull { source ->
            val form = SwipeWordNormalizer.normalize(source.word) ?: return@mapNotNull null
            val word = form.display
            val geometry = form.geometry
            if (
                !seen.add("${source.languageTag ?: "*"}|$word")
            ) {
                return@mapNotNull null
            }

            val centers = mutableListOf<SwipePoint>()
            var previous: Char? = null
            geometry.forEach { letter ->
                if (letter != previous) {
                    centers += QwertyKeyLayout.center(letter) ?: return@mapNotNull null
                }
                previous = letter
            }
            val first = centers.firstOrNull() ?: return@mapNotNull null
            val last = centers.last()

            val repeatedLetters = mutableMapOf<Char, Int>()
            previous = null
            geometry.forEach { letter ->
                if (letter == previous) {
                    repeatedLetters[letter] = repeatedLetters.getOrDefault(letter, 0) + 1
                }
                previous = letter
            }

            Entry(
                word = word,
                geometry = geometry,
                rank = source.rank,
                languageTag = source.languageTag,
                centers = centers,
                repeatedLetters = repeatedLetters,
                pathLength = length(centers),
                firstCenter = first,
                lastCenter = last,
            )
        }
    }

    fun candidates(
        trace: String,
        limit: Int = 3,
        previousWord: String? = null,
    ): List<String> = candidates(
        path = trace.mapNotNull(QwertyKeyLayout::center),
        limit = limit,
        previousWord = previousWord,
    )

    fun candidates(
        path: List<SwipePoint>,
        limit: Int = 3,
        previousWord: String? = null,
    ): List<String> = scoredCandidates(
        path = path,
        repeatedLetters = null,
        limit = limit,
        previousWord = previousWord,
        languageTag = null,
    )
        .filter { it.score <= REJECTION_SCORE }
        .map(ScoredEntry::word)

    /**
     * Rich timed recognition Interface. Candidate confidences and the top-two
     * margin are returned even when recognition conservatively abstains.
     */
    fun recognize(
        samples: List<SwipePathSample>,
        limit: Int = 3,
        previousWord: String? = null,
        languageTag: String? = null,
    ): SwipeRecognitionResult {
        if (samples.size < 2 || limit <= 0) {
            return SwipeRecognitionResult.abstaining(SwipeAbstentionReason.TOO_SHORT)
        }
        if (!valid(samples)) {
            return SwipeRecognitionResult.abstaining(SwipeAbstentionReason.INVALID_SAMPLES)
        }

        val scored = scoredCandidates(
            path = samples.map(SwipePathSample::point),
            repeatedLetters = repeatedLetterEvidence(samples, dwellConfiguration),
            limit = limit,
            previousWord = previousWord,
            languageTag = languageTag,
        ).filter { it.score <= RECOGNITION_SCORE_LIMIT }
        if (scored.isEmpty()) {
            return SwipeRecognitionResult.abstaining(SwipeAbstentionReason.NO_CANDIDATE)
        }

        val candidates = scored.map { candidate ->
            SwipeCandidate(candidate.word, confidence(candidate.score))
        }
        val confidence = candidates.first().confidence
        val margin = if (candidates.size > 1) {
            max(0.0, confidence - candidates[1].confidence)
        } else {
            confidence
        }
        val reason = when {
            confidence < MINIMUM_CONFIDENCE -> SwipeAbstentionReason.LOW_CONFIDENCE
            margin < MINIMUM_MARGIN -> SwipeAbstentionReason.AMBIGUOUS
            else -> null
        }
        return SwipeRecognitionResult(
            candidates = candidates,
            acceptedCandidate = candidates.firstOrNull().takeIf { reason == null },
            confidence = confidence,
            margin = margin,
            abstentionReason = reason,
        )
    }

    private fun scoredCandidates(
        path: List<SwipePoint>,
        repeatedLetters: Map<Char, Int>?,
        limit: Int,
        previousWord: String?,
        languageTag: String?,
    ): List<ScoredEntry> {
        if (path.size < 2 || limit <= 0 || entries.isEmpty()) return emptyList()
        val start = path.first()
        val end = path.last()
        val requestedLanguage = languageTag?.let(::baseLanguage)
        val eligibleEntries = entries.filter { entry ->
            requestedLanguage == null ||
                entry.languageTag == null ||
                entry.languageTag == requestedLanguage
        }
        val sampledPath = resample(path, SAMPLE_COUNT)
        val sampledShape = shapeNormalized(sampledPath)
        val pathLength = length(path)
        val vocabularySize = max(eligibleEntries.size, 1)
        val continuations = SuggestionEngine.commonContinuations(previousWord)
        val bestByWord = mutableMapOf<String, Double>()

        eligibleEntries.forEach { entry ->
            val startDistance = distance(entry.firstCenter, start)
            val endDistance = distance(entry.lastCenter, end)
            if (
                startDistance > ANCHOR_TOLERANCE ||
                endDistance > ANCHOR_TOLERANCE ||
                abs(ln((entry.pathLength + 0.5) / (pathLength + 0.5))) > ln(2.3)
            ) {
                return@forEach
            }

            val idealPath = resample(entry.centers, SAMPLE_COUNT)
            val location = meanDistance(sampledPath, idealPath) / 3.0
            val shape = meanDistance(sampledShape, shapeNormalized(idealPath)) * 2.0
            val rankScore = ln(1.0 + entry.rank) / ln(1.0 + vocabularySize)
            var score = 0.40 * location.coerceAtMost(1.5) +
                0.32 * shape.coerceAtMost(1.5) +
                0.18 * rankScore +
                0.05 * (startDistance + endDistance)
            if (repeatedLetters != null) {
                score += repeatedLetterScore(
                    expected = entry.repeatedLetters,
                    observed = repeatedLetters,
                )
            }
            if (entry.word in continuations) score -= 0.10
            bestByWord[entry.word] = min(bestByWord[entry.word] ?: Double.POSITIVE_INFINITY, score)
        }

        return bestByWord.map { (word, score) -> ScoredEntry(word, score) }
            .sortedWith(compareBy<ScoredEntry> { it.score }.thenBy { it.word })
            .take(limit)
    }

    private fun repeatedLetterEvidence(
        samples: List<SwipePathSample>,
        configuration: SwipeDwellConfiguration,
    ): Map<Char, Int> {
        val runs = mutableListOf<KeyRun>()
        samples.forEach { sample ->
            val key = QwertyKeyLayout.nearestKey(sample.point) ?: return@forEach
            if (runs.lastOrNull()?.key == key) {
                runs.last().samples += sample
            } else {
                runs += KeyRun(key, mutableListOf(sample))
            }
        }

        val evidence = mutableMapOf<Char, Int>()
        runs.forEach { run ->
            val first = run.samples.firstOrNull() ?: return@forEach
            val last = run.samples.last()
            if (
                run.samples.size < configuration.minimumSamples ||
                last.timestampMilliseconds - first.timestampMilliseconds < configuration.minimumMilliseconds
            ) {
                return@forEach
            }
            val maximumDrift = run.samples.maxOf { sample ->
                hypot(sample.x - first.x, sample.y - first.y)
            }
            if (maximumDrift > configuration.maximumDriftKeyUnits) return@forEach
            evidence[run.key] = evidence.getOrDefault(run.key, 0) + 1
        }
        return evidence
    }

    private fun repeatedLetterScore(
        expected: Map<Char, Int>,
        observed: Map<Char, Int>,
    ): Double = (expected.keys + observed.keys).sumOf { key ->
        val expectedCount = expected.getOrDefault(key, 0)
        val observedCount = observed.getOrDefault(key, 0)
        val matched = min(expectedCount, observedCount)
        val missing = max(0, expectedCount - observedCount)
        val unexpected = max(0, observedCount - expectedCount)
        -0.12 * matched + 0.18 * missing + 0.16 * unexpected
    }

    private fun confidence(score: Double): Double =
        (1.0 / (1.0 + exp(6.0 * (score - 0.45)))).coerceIn(0.0, 1.0)

    private fun baseLanguage(languageTag: String): String = languageTag
        .lowercase()
        .split('-', '_')
        .firstOrNull()
        ?: languageTag.lowercase()

    private fun valid(samples: List<SwipePathSample>): Boolean {
        var previousTimestamp = Double.NEGATIVE_INFINITY
        return samples.all { sample ->
            val valid = sample.x.isFinite() &&
                sample.y.isFinite() &&
                sample.timestampMilliseconds.isFinite() &&
                sample.timestampMilliseconds >= previousTimestamp
            previousTimestamp = sample.timestampMilliseconds
            valid
        }
    }

    private companion object {
        const val SAMPLE_COUNT = 32
        const val ANCHOR_TOLERANCE = 1.6
        const val REJECTION_SCORE = 0.62
        const val RECOGNITION_SCORE_LIMIT = 0.78
        const val MINIMUM_CONFIDENCE = 0.50
        const val MINIMUM_MARGIN = 0.03

        fun distance(left: SwipePoint, right: SwipePoint): Double =
            hypot(left.x - right.x, left.y - right.y)

        fun length(points: List<SwipePoint>): Double = points.zipWithNext()
            .sumOf { (left, right) -> distance(left, right) }

        fun resample(points: List<SwipePoint>, count: Int): List<SwipePoint> {
            val first = points.firstOrNull() ?: return emptyList()
            val total = length(points)
            if (points.size < 2 || total <= 0.0) return List(count) { first }

            val interval = total / (count - 1)
            val result = mutableListOf(first)
            var accumulated = 0.0
            var previous = first
            for (point in points.drop(1)) {
                var segment = distance(previous, point)
                var segmentStart = previous
                while (accumulated + segment >= interval && result.size < count) {
                    val ratio = (interval - accumulated) / segment
                    val sample = SwipePoint(
                        x = segmentStart.x + ratio * (point.x - segmentStart.x),
                        y = segmentStart.y + ratio * (point.y - segmentStart.y),
                    )
                    result += sample
                    segment = accumulated + segment - interval
                    accumulated = 0.0
                    segmentStart = sample
                }
                accumulated += segment
                previous = point
            }
            while (result.size < count) result += points.last()
            return result
        }

        fun shapeNormalized(points: List<SwipePoint>): List<SwipePoint> {
            if (points.isEmpty()) return points
            val centroid = SwipePoint(
                x = points.sumOf(SwipePoint::x) / points.size,
                y = points.sumOf(SwipePoint::y) / points.size,
            )
            val scale = max(
                max(
                    points.maxOf(SwipePoint::x) - points.minOf(SwipePoint::x),
                    points.maxOf(SwipePoint::y) - points.minOf(SwipePoint::y),
                ),
                0.01,
            )
            return points.map { point ->
                SwipePoint((point.x - centroid.x) / scale, (point.y - centroid.y) / scale)
            }
        }

        fun meanDistance(left: List<SwipePoint>, right: List<SwipePoint>): Double {
            if (left.size != right.size || left.isEmpty()) return Double.POSITIVE_INFINITY
            return left.indices.sumOf { distance(left[it], right[it]) } / left.size
        }
    }
}
